### PR TITLE
implement hibernation

### PR DESCRIPTION
**Pull request by harshil1712**

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,65 +1,120 @@
...
- * - Run `npm run deploy` to publish your application
- *
- * Bind resources to your worker in `wrangler.toml`. After adding bindings, a type definition for the
- * `Env` object can be regenerated with `npm run cf-typegen`.
- *
- * Learn more at https://developers.cloudflare.com/durable-objects
- */
-
-/** A Durable Object's behavior is defined in an exported Javascript class */
+// Use npm run cf-typegen to generate the type definitions for the Durable Object
```
Comment (joshthoward) [inline src/index.ts]: Is this actually needed? Installing with @cloudflare/workers-types and wrangler as a dependency is usually sufficient.

File: src/index.ts [hunk 2]
```diff
@@ -1,65 +1,120 @@
...
+
+// Worker
+export default {
+	async fetch(request: Request, env: Env, ctx: ExecutionContext): Promise<Response> {
+		if (request.url.endsWith('/websocket')) {
+			// Expect to receive a WebSocket Upgrade request.
+			// If there is one, accept the request and return a WebSocket Response.
+			const upgradeHeader = request.headers.get('Upgrade');
```
Comment (joshthoward) [inline src/index.ts:9]: We should also check that this is an HTTP GET and if not return 400.

File: src/index.ts [hunk 3]
```diff
@@ -1,65 +1,120 @@
...
+		// (run the `constructor`) and deliver the message to the appropriate handler.
+		this.ctx.acceptWebSocket(server);
+
+		// Generate a random UUID for the session.
+		const id = crypto.randomUUID();
+
+		// Attach the session ID to the WebSocket connection and serialize it.
+		// This is necessary to restore the state of the connection when the Durable Object wakes up.
+		server.serializeAttachment({
+			...server.deserializeAttachment(),
```
Comment (joshthoward) [inline src/index.ts]: Maybe I'm missing something, but I cannot think of a case where server would have an attachment at this stage. So I don't believe that this line is needed.

File: src/index.ts [hunk 4]
```diff
@@ -1,65 +1,120 @@
...
+			statusText: 'Bad Request',
+			headers: {
+				'Content-Type': 'text/plain',
+			},
+		});
+	},
+};
+
+// Durable Object
 export class WebSocketHibernationServer extends DurableObject {
```
Comment (joshthoward) [inline src/index.ts:44]: I would make this name more descriptive of a topology. `WebSocketBroadcast` seems more informative. I think it would be really helpful to also include an example of this using the standard WebSocket libraries.

File: src/index.ts [hunk 5]
```diff
@@ -1,65 +1,120 @@
...
+		// Add the WebSocket connection to the map of active sessions.
+		this.sessions.set(server, { id });
+
+		return new Response(null, {
+			status: 101,
+			webSocket: client,
+		});
+	}
+
+	async webSocketMessage(sender: WebSocket, message: ArrayBuffer | string) {
```
Comment (joshthoward) [inline src/index.ts]: It would be nice to show off the auto response feature https://developers.cloudflare.com/durable-objects/api/state/#setwebsocketautoresponse. Maybe something like zero cost websocket pings.

**Review by joshthoward — APPROVED** [review]
Looks good overall, we can merge once these suggestions are addressed.

File: src/index.ts (hunk 1 above)
Comment (harshil1712) [reply src/index.ts]: Ah, I didn't know about that. Our docs don't have a lot of examples with `@cloudflare/workers-types`

File: src/index.ts (hunk 3 above)
Comment (harshil1712) [reply src/index.ts]: I was trying to address this comment:  https://github.com/cloudflare/cloudflare-docs/pull/17670/files#r1808955786

if this isn't the correct approach, where should the ID be generated and attached to the connection?

File: src/index.ts (hunk 4 above)
Comment (harshil1712) [reply src/index.ts:44]: The main idea is to show different aspects of the Hibernation API. Broadcasting message is more on the educational side. I like to think of it like an end-to-end example that folks can copy and paste if they want to have hibernated WebSockets. Based on the application, you can broadcast message, or not.

Yes, I'll be updating the standard WebSocket example once this is merged. I believe it would be simpler in terms of complexity and won't take much time.

File: src/index.ts (hunk 3 above)
Comment (harshil1712) [reply src/index.ts]: Going through the chat example, it looks like we are doing the same there, but in the `handleSession` method.
https://github.com/cloudflare/workers-chat-demo/blob/97bacfa773c673ef35f974dc076ba65c42fede71/src/chat.mjs#L303

File: src/index.ts [hunk 6]
```diff
@@ -1,65 +1,127 @@
...
-/** A Durable Object's behavior is defined in an exported Javascript class */
+// Worker
+export default {
+	async fetch(request: Request, env: Env, ctx: ExecutionContext): Promise<Response> {
+		if (request.url.endsWith('/websocket')) {
+			// Expect to receive a WebSocket Upgrade request.
+			// If there is one, accept the request and return a WebSocket Response.
+			const upgradeHeader = request.headers.get('Upgrade');
+			if (!upgradeHeader || upgradeHeader !== 'websocket') {
+				return new Response('Durable Object expected Upgrade: websocket', {
```
Comment (cacheburner) [inline src/index.ts]: It seems a bit awkward to say `Durable Object expected FOO` from within the Worker code. 

Let's say I'm someone with very little experience with Workers, I might wonder if we're in the DO at this point (yes there's a comment above `export default` at the top, but still).

I'm not sure how I'd rephrase it, or if we even should, but maybe `Durable Object expects FOO`, or `Worker expected FOO`?

File: src/index.ts [hunk 7]
```diff
@@ -1,65 +1,127 @@
...
+				});
+			}
+
+			if (request.method !== 'GET') {
+				return new Response('Durable Object expected GET method', {
+					status: 400,
+				});
+			}
+
+			// This example will refer to the same Durable Object,
```
Comment (cacheburner) [inline src/index.ts]: Could we rephrase this comment? Something like "_Since we are hard coding the Durable Object ID by providing the constant name 'foo', all requests to this Worker will be sent to the same Durable Object instance_".

File: src/index.ts [hunk 8]
```diff
@@ -1,65 +1,127 @@
...
+			// since the name "foo" is hardcoded.
+			let id = env.WEBSOCKET_HIBERNATION_SERVER.idFromName('foo');
+			let stub = env.WEBSOCKET_HIBERNATION_SERVER.get(id);
+
+			return stub.fetch(request);
+		}
+
+		return new Response(null, {
```
Comment (cacheburner) [inline src/index.ts]: If the intention is to have people clone and run this example, I'd suggest having the default path return usage :slightly_smiling_face:. 

For example:

```
Supported endpoints:

/websocket
    - Expects a WebSocket upgrade request
``` 
I've previously [deployed a DO](https://typescript.milanmiladinovic.workers.dev/) that does this.

This way, when sending an initial fetch, they won't be greeted with a `Bad Request` (which might make them think the deployment isn't working).

File: src/index.ts [hunk 9]
```diff
@@ -1,65 +1,127 @@
...
+// Durable Object
 export class WebSocketHibernationServer extends DurableObject {
-	/**
-	 * The constructor is invoked once upon creation of the Durable Object, i.e. the first call to
-	 * 	`DurableObjectStub::get` for a given identifier (no-op constructors can be omitted)
-	 *
-	 * @param ctx - The interface for interacting with Durable Object state
-	 * @param env - The interface to reference bindings declared in wrangler.toml
-	 */
+	// Keep track of all WebSocket connections
```
Comment (cacheburner) [inline src/index.ts]: I would expand on this state variable.

It would be really helpful to explain how it retains the `Map` while the DO is in memory, but when we "hibernate" we lose that state and need to reconstruct it in the `constructor()`.

File: src/index.ts [hunk 10]
```diff
@@ -1,65 +1,127 @@
...
+	sessions: Map<WebSocket, any>;
+
 	constructor(ctx: DurableObjectState, env: Env) {
 		super(ctx, env);
+		this.sessions = new Map();
+
+		// DO saves the connection information when it goes to hibernation
```
Comment (cacheburner) [inline src/index.ts]: I'd drop the comment on line 49: `DO saves the connection information when it goes to hibernation`. It doesn't explain what's happening, and mostly applies to the `serializeAttachment` on line 86 below.

It might be easier to just clarify something like:
```
// As part of constructing our Durable Object instance, 
// we want to wake up any hibernating WebSockets and 
// place them back in our `sessions` map.
```

File: src/index.ts [hunk 11]
```diff
@@ -1,65 +1,127 @@
...
+		// Get all WebSocket connections from the DO
+		this.ctx.getWebSockets().forEach((ws) => {
+			// Checks if this is a new DO
```
Comment (cacheburner) [inline src/index.ts]: What does `Checks if this is a new DO` mean? This should probably be deleted.

Also, rather than calling `deserializeAttachment()` twice (which actually does result in v8 deserialization), it would be best if we did something like:

```
let attachment = ws.deserializeAttachment();

// If we had previously attached state to our WebSocket, let's add it to our
// `sessions` map to restore the state of the connection.
if (attachment) {
    const { ...session } = attachment;
    this.sessions.set(ws, { ...session});
}
```

I'm not sure if the `const { ...session } = attachment;` line is necessary, I'm not a JS programmer :sweat_smile:

File: src/index.ts [hunk 12]
```diff
@@ -1,65 +1,127 @@
...
+			if (ws.deserializeAttachment()) {
+				// If a DO is waking up, deserialize the attachment.
```
Comment (cacheburner) [inline src/index.ts]: I figure this will change because of my other comment above, but `If a DO is waking up` doesn't make sense in this context. We _know_ the DO is starting up because we're in the `constructor()`.

File: src/index.ts [hunk 13]
```diff
@@ -1,65 +1,127 @@
...
+				// This helps to restore the state of the connection
+				const { ...session } = ws.deserializeAttachment();
+				this.sessions.set(ws, { ...session });
+			}
+		});
+
+		// Sets an application level auto response that does not wake hibernated WebSockets.
```
Comment (cacheburner) [inline src/index.ts:68]: ~~Sets~~ Set an application....

Comment (cacheburner) [reply src/index.ts:68]: It would be great to expand on this comment btw!

I think customers get tripped up by this feature quite easily, but once they understand it they appreciate it a lot.

```
// Setting this auto-response means that whenever any of the WebSockets
// associated with this Durable Object receive the message `ping`, they will 
// automatically respond with `pong` without delivering the event to JavaScript.
//
// This means the message `ping` will not wake your Durable Object if it is hibernating,
// and it also does not incur a charge!
//
// Setting an auto-response also ensures the connection stays alive. 
// If no messages are sent for a while, a TCP connection timeout on Cloudflare's 
// servers may close the connection.
```

File: src/index.ts [hunk 14]
```diff
@@ -1,65 +1,127 @@
...
-		return `Hello, ${name}!`;
+	async fetch(request: Request): Promise<Response> {
+		// Creates two ends of a WebSocket connection.
+		const webSocketPair = new WebSocketPair();
+		const [client, server] = Object.values(webSocketPair);
+
+		// Calling `acceptWebSocket()` informs the runtime that this WebSocket is to begin terminating
+		// request within the Durable Object. It has the effect of "accepting" the connection,
+		// and allowing the WebSocket to send and receive messages.
+		// Unlike `ws.accept()`, `state.acceptWebSocket(ws)` informs the Workers Runtime that the WebSocket
```
Comment (cacheburner) [inline src/index.ts]: `state.acceptWebSocket` doesn't make sense when we then write `this.ctx.acceptWebSocket()` below :wink:.

File: src/index.ts (hunk 3 above)
Comment (cacheburner) [reply src/index.ts]: @harshil1712 I actually have the context for doing that [here](https://github.com/cloudflare/workers-chat-demo/pull/32#discussion_r1198304406).

The concern is basically just if someone were to add more stuff to the attachment and call `serializeAttachment()` multiple times, it could break.

I don't think it's too important here though, we can just do `server.serializeAttachment({id});`

File: src/index.ts (hunk 9 above)
Comment (cacheburner) [reply src/index.ts]: Also, why is the value type `any`? Could we make it an explicit type instead? I think that would make things clearer.

File: src/index.ts [hunk 15]
```diff
@@ -1,65 +1,127 @@
...
-		let stub = env.WEBSOCKET_HIBERNATION_SERVER.get(id);
-
-		// We call the `sayHello()` RPC method on the stub to invoke the method on the remote
-		// Durable Object instance
-		let greeting = await stub.sayHello('world');
-
-		return new Response(greeting);
-	},
-} satisfies ExportedHandler<Env>;
+	async webSocketMessage(sender: WebSocket, message: ArrayBuffer | string) {
```
Comment (cacheburner) [inline src/index.ts]: nit: I don't think `sender` makes sense here, this is actually the `receiver` of a message from a client.

Either way, why make the distinction? `ws` should be fine (that is what it is, after all).

File: src/index.ts [hunk 16]
```diff
@@ -1,65 +1,127 @@
...
+		// Get the session associated with the WebSocket connection.
+		const session = this.sessions.get(sender);
+
+		// Upon receiving a message from the client, the server replies with the same message, the session ID of the sender,
```
Comment (cacheburner) [inline src/index.ts]: nit: ~~session ID of the sender~~ --> "_session ID of the connection_".

The connection's ID doesn't change for the duration of the connection.

File: src/index.ts [hunk 17]
```diff
@@ -1,65 +1,127 @@
...
+		// and the total number of connections with the "[Durable Object]: " prefix
+		sender.send(`[Durable Object] message: ${message}, from: ${session.id}. Total connections: ${this.ctx.getWebSockets().length}`);
+
+		// Send a message to all WebSocket connections, loop over all the connected WebSockets.
+		// This wakes up all the connections and the DO is no longer in hibernation.
```
Comment (cacheburner) [inline src/index.ts]: `This wakes up all the connections and the DO is no longer in hibernation.`.

Well... not exactly.

1. If the DO was hibernating prior to receiving this WebSocket message, it is no longer hibernating now because we would've called the `constructor()` _first_, then called `webSocketMessage()`.
2. Inside of you `constructor()`, you call `this.ctx.getWebSockets()` to rebuild the `sessions` map. In doing this, you already recreate all the JS WebSockets.
3. Technically, you don't even need to use `this.ctx.getWebSockets()` here, because you could just use your `sessions` map. That might actually be even better, because then you could also include each connection's session ID in the `ws.send()`.

File: src/index.ts [hunk 18]
```diff
@@ -1,65 +1,127 @@
...
+		this.ctx.getWebSockets().forEach((ws) => {
+			ws.send(`[Durable Object] message: ${message}, from: ${session.id}. Total connections: ${this.ctx.getWebSockets().length}`);
+		});
+
+		// Send a message to all WebSocket connections except the sender, loop over all the connected WebSockets and filter out the sender.
+		// This wakes up all the connections and the DO is no longer in hibernation.
```
Comment (cacheburner) [inline src/index.ts]: See comment above about DO being in hibernation.

**Comment by harshil1712** [comment]
Hey @joshthoward and @MellowYarker, thank you for the reviews! I have made the changes, and I think it is good to go. I'll update the example on the docs PR
